### PR TITLE
Unet model contribution

### PR DESCRIPTION
**Pull request by pctablet505**
This pull request adds support for the UNet architecture for image segmentation to the codebase. It introduces both a reusable UNet backbone model and a UNet-based image segmenter task model, along with comprehensive tests for both. The new models are registered for import and integrated into the package API.

**New UNet architecture and segmentation support:**

* Added `UNetBackbone`, a configurable Keras implementation of the UNet encoder-decoder architecture with skip connections, supporting dynamic input shapes, batch normalization, and dropout.
* Added `UNetImageSegmenter`, a segmentation task model that wraps a `UNetBackbone` and outputs segmentation masks for a configurable number of classes, with optional output activation.

**Testing and quality assurance:**

* Added `UNetBackboneTest` to verify model construction, dynamic input handling, batch norm, dropout, different depths, model saving, and preset loading.
* Added `UNetImageSegmenterTest` to check segmenter basics, dynamic input shapes, varying class counts, activation options, and model saving.

**API and import integration:**

* Registered `UNetBackbone` and `UNetImageSegmenter` in the package-level API (`keras_hub/api/models/__init__.py` and `keras_hub/src/models/unet/__init__.py`). [[1]](diffhunk://#diff-75218135d0afa7f249eeb39820394eef664397ad12b220ded2dce729b038fa9eR750-R753) [[2]](diffhunk://#diff-b1902473b32a7ac83d3a88bf75aaaf67f994de62062ffb0373cb2398b0924888R1-R2)## Description of the change

## Checklist

- [ ] I have added all the necessary unit tests for my change.
- [ ] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [ ] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [ ] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [ ] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [ ] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by pctablet505** [comment]
@gemini-code-assist review

File: keras_hub/src/models/unet/unet_encoder.py [hunk 1]
```diff
@@ -0,0 +1,404 @@
+"""UNet Encoder implementation."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_encoder.py]: We don't do file specific docstring, it's limited to class and functions level.

File: keras_hub/src/models/unet/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from keras_hub.src.models.unet.unet_backbone import UNetBackbone
+from keras_hub.src.models.unet.unet_decoder import UNetDecoder
+from keras_hub.src.models.unet.unet_encoder import UNetEncoder
+from keras_hub.src.models.unet.unet_image_segmenter import UNetImageSegmenter
+from keras_hub.src.models.unet.unet_image_segmenter import (
+    UNetImageSegmenterPreprocessor,
+)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/__init__.py:7]: You can just import UNetBackbone, which is required for registering presets if we planning to add any pre-trained model weights.

File: keras_hub/src/models/unet/unet_backbone.py [hunk 3]
```diff
@@ -0,0 +1,850 @@
...
+        self.use_residual = use_residual
+        self.use_attention = use_attention
+        self.kernel_initializer = kernel_initializer
+        self._backbone_feature_names = None
+
+        # === Functional Model ===
+        inputs = keras.layers.Input(shape=image_shape)
+
+        # Build encoder
+        bottleneck, encoder_outputs = self._build_encoder(
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_backbone.py]: Instead of re implementing the encoder, can't we just import and use the unet_encoder here, that way it would be much fewer code and less confusion to trace any error.

File: keras_hub/src/models/unet/unet_backbone.py [hunk 4]
```diff
@@ -0,0 +1,850 @@
...
+            filters=filters,
+            use_batch_norm=use_batch_norm,
+            use_residual=use_residual,
+            kernel_initializer=kernel_initializer,
+            data_format=data_format,
+            dtype=dtype,
+        )
+
+        # Build decoder
+        x = self._build_decoder(
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_backbone.py]: Same for this, we can just use unet_decoder, current implementation just has standalone Encoder and Decoder implementation and just exposes UNetEncoder, UNetDecoder, instead reusing would be better.

File: keras_hub/src/models/unet/unet_backbone_test.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+class UNetBackboneTest(TestCase):
+    def setUp(self):
+        self.init_kwargs = {
+            "depth": 3,
+            "filters": 32,
+            "image_shape": (None, None, 3),
+            "data_format": "channels_last",
+        }
+        self.input_size = 128
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_backbone_test.py:20]: Can you test with the smallest setup possible to keep the test lightweight.

File: keras_hub/src/models/unet/unet_backbone_test.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        model = UNetBackbone(**self.init_kwargs)
+
+        # Test with different input sizes
+        input_128 = ops.ones((1, 128, 128, 3))
+        output_128 = model(input_128)
+        self.assertEqual(output_128.shape, (1, 128, 128, 32))
+
+        input_256 = ops.ones((1, 256, 256, 3))
+        output_256 = model(input_256)
+        self.assertEqual(output_256.shape, (1, 256, 256, 32))
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_backbone_test.py:65]: Smallest setup should be fine, may be 32 and 16?

File: keras_hub/src/models/unet/unet_decoder.py [hunk 7]
```diff
@@ -0,0 +1,470 @@
+"""UNet Decoder implementation."""
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_decoder.py]: Remove this line.

File: keras_hub/src/models/unet/unet_image_segmenter.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    This preprocessor simply passes through the input images and labels
+    without any modification, since UNet can handle variable input sizes.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(image_converter=None, **kwargs)
+
+    def call(self, x, y=None, sample_weight=None):
+        # For UNet, we don't need any preprocessing - just pass through
+        return keras.utils.pack_x_y_sample_weight(x, y, sample_weight)
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_image_segmenter.py]: You can create this in a separate file.
I think you can also pass `image_converter_cls = None` which will give the packed data without preprocessing.

File: keras_hub/src/models/unet/unet_image_segmenter.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        num_classes: int. The number of classes for the segmentation model.
+            Note that the `num_classes` contains the background class, and the
+            classes from the data should be represented by integers with range
+            `[0, num_classes)`.
+        activation: str or callable. The activation function to use on the
+            output layer. Set `activation=None` to return the output logits.
+            Defaults to `"softmax"`.
+        preprocessor: A preprocessor instance or `None`. If `None`, this model
+            will not apply preprocessing, and inputs should be preprocessed
+            before calling the model.
```
Comment (sachinprasadhs) [inline keras_hub/src/models/unet/unet_image_segmenter.py]: This is ambiguous, since you have stated no preprcessing required in the preprocessor logic and this statement is not matching.

File: keras_hub/src/tests/test_case.py [hunk 10]
```diff
@@ -991,6 +991,9 @@ def run_vision_backbone_test(
                 init_kwargs["image_shape"] = tuple(
                     reversed(init_kwargs["image_shape"])
                 )
+            if "data_format" in init_kwargs:
+                init_kwargs = init_kwargs.copy()
+                init_kwargs["data_format"] = "channels_first"
```
Comment (sachinprasadhs) [inline keras_hub/src/tests/test_case.py:996]: In the existing above logic in this file, test already performs channel_first format test by converting the channel last to channel first, you can just use `run_vision_backbone_test`.

**Review by sachinprasadhs — COMMENTED** [review]
Thanks, I have done one pass and made some comments.
Also, check Gemini comments, if those are valid you can address it, else you can mark them as resolved.
Also, is this implemtation is limited to architecture only or are we adding any checkpoints to kaggle?

**Comment by pctablet505** [comment]
@gemini-code-assist review

**Comment by pctablet505** [comment]
Model not required
